### DISCUSSION TITLE
Query Regarding Specifying Time Steps (W) for Viscosity Measurement Using hoomd.md.update.ReversePerturbationFlow

### DISCUSSION THREAD

**Discussion by kojikokojiko** [glotzerlab/hoomd-blue#1682] [Support] [answered]
Dear hoomd-users Group,

I am currently utilizing the hoomd.md.update.ReversePerturbationFlow for calculating viscosity through the Mueller-Plathe method. My inquiry pertains to the specification of the number of time steps (W) until momentum exchange, as mentioned in the original research papers.

In my current understanding and application, I haven't found a way to directly specify W for measuring viscosity. This specification seems crucial for replicating the procedures outlined in the original literature and for ensuring the accuracy of the viscosity measurements.

Could you please advise if there is a method or parameter within HOOMD-blue that allows for the specification of W during the viscosity measurement process? Any guidance or suggestions for approximating this functionality, if not directly available, would also be greatly appreciated.

Thank you for your time and assistance.

**Comment by mphoward** [comment, marked as answer]
I am not the author of this code, but I am familiar with the method. The implementation of the reverse perturbation shearing method in HOOMD is **not** the standard algorithm described by Müller-Plathe. The reason is that the `flow_target` parameter is the integrated (total) momentum flux, and swaps are attempted to keep pace with this target as the simulation runs (using a control-like scheme). Hence, you cannot specify the number of steps between swaps or the number of swaps that will be attempted, although this is how the algorithm normally works. I think the original author had some research motivations for doing things this way rather than the standard approach.

We implemented the standard algorithm in a plugin:

https://azplugins.readthedocs.io/en/latest/api/flow.html#azplugins.flow.reverse_perturbation

However, note that this plugin is currently stuck at HOOMD 2, and I am not sure when there will be enough time to port it to HOOMD 4.
